File: spring-web-java/src/main/java/spring/web/java/application/service/RequestLockService.java
package spring.web.java.application.service;

import org.springframework.stereotype.Service;

import lombok.RequiredArgsConstructor;
import spring.web.java.domain.repository.RequestLockRepository;
import spring.web.java.presentation.exception.DuplicateRequestException;

@Service
@RequiredArgsConstructor
public class RequestLockService {

	private final RequestLockRepository requestLockRepository;

	public void validate(Long memberId, String method, String uri) {
		if (!requestLockRepository.setIfAbsent(memberId, method, uri)) {
			throw new DuplicateRequestException(memberId, method, uri);
		}
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/application/service/RequestLockService.kt
package spring.web.kotlin.application.service

import org.springframework.stereotype.Service
import spring.web.kotlin.domain.repository.RequestLockRepository
import spring.web.kotlin.presentation.exception.DuplicateRequestException

@Service
class RequestLockService(
    private val requestLockRepository: RequestLockRepository
) {
    fun validate(memberId: Long, method: String, uri: String) {
        if (!requestLockRepository.setIfAbsent(memberId, method, uri)) {
            throw DuplicateRequestException(memberId, method, uri)
        }
    }
}